src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -5963,51 +5963,94 @@ void MacroAssembler::leave() {
 // For more details on PAC see pauth_aarch64.hpp.
 
 // Sign the LR. Use during construction of a stack frame, before storing the LR to memory.
-// Uses the FP as the modifier.
+// Uses relative SP as the modifier.
+// Note that relative SP = signing SP - initial SP, where
+//   signing SP is the lowest address of the frame record where LR is saved, and
+//   initial SP is thread->last_continuation()->entry_sp().
+//
+//                  |          |
+//                  |==========|
+//                  |          |  // Continuation.run
+//                  |          |
+//                  |==========|\
+//                  |          | | ContinuationEntry          Address
+//                  |          | |                             high
+//  initial SP ---> |==========|/                                |
+//                  |          |                                 |
+//                  | ...      |                                \|/
+//                  |          |                                low
+//                  |==========|\
+//                  |  LR      | |
+//                  |  FP      | | frame record
+//  signing SP ---> |----------|/
+//                  |          |
+//                  |          |
 //
-void MacroAssembler::protect_return_address() {
+// Note that initial SP is zero for non-virtual thread.
+//
+void MacroAssembler::protect_return_address(Register thread) {
   if (VM_Version::use_rop_protection()) {
     check_return_address();
-    // The standard convention for C code is to use paciasp, which uses SP as the modifier. This
-    // works because in C code, FP and SP match on function entry. In the JDK, SP and FP may not
-    // match, so instead explicitly use the FP.
-    pacia(lr, rfp);
+    if (thread == noreg) {
+      paciasp();
+    } else {
+      stp(rscratch1, zr, Address(pre(sp, -2 * wordSize)));
+      ldr(rscratch1, Address(thread, JavaThread::cont_entry_offset()));
+      sub(rscratch1, sp, rscratch1);
+      pacia(lr, rscratch1);
+      ldp(rscratch1, zr, Address(post(sp, 2 * wordSize)));
+    }
```

Comment:
```suggestion
      ldr(lr, Address(thread, JavaThread::cont_entry_offset()));
      sub(lr, sp, lr);
      pacia(lr, lr);
    }
```
Maybe?

Reply:
I'm afraid not.
In this way, the original value of `lr` would be overwritten before `pacia` instruction, and we turn to protect `relative SP` with `relative SP` as the modifier, rather than protecting `lr` with `relative SP` as the modifier.

We may get some basics of `ARM pauth` from this slides, https://events.static.linuxfound.org/sites/events/files/slides/slides_23.pdf

For simplicity, we can view the process of "PAC signing" and "PAC authentication" as the **generation** and **verification** of the **PAC code** (something like one hash code).

For PAC generation, three inputs are taken in
- the pointer to protection, i.e. LR in our context
- one key, which is maintained by the underlying kernel
- context, i.e. the modifier. we can use any value in theory, like constants, zero, some runtime values(SP, FP, thread id etc)

In our context, after `pacia` instruction, the LR under protection would turn to be "PAC code + LR", where the generated PAC code is put at the higher N bits and the lower 48 bit is still the original LR.
Hence, the PACed LR cannot be accessed directly, until the corresponding authentication is conducted.